README.md
# python-powerservice

CHANGELOG.rst

Changelog
=========

0.0.0 (2021-06-24)
------------------




tests/test_powerservice.py
""" Tests for the powerservice"""
import pytest
from powerservice.trading import check_if_valid_date, generate_new_random_trade_position

cases = [("", False),
         ("01/04/2015", True),
         ("30/04/2015", True),
         (1, False),
         ("01/30/2015", False),
         ("30/04/bla", False)
         ]


@pytest.mark.parametrize("date,expected", cases)
def test_date_checker(date, expected):
    """ Check that only d/m/y formatted date is accepted"""
    assert check_if_valid_date(date) == expected

#
# def test_generate_new_random_trade_position_period():
#     """ Try to generate a new random trade position"""
#     new_trade = generate_new_random_trade_position(date="01/04/2015")
#     period_list = new_trade["period"]
#
#     assert period_list[0] == 1 and period_list[-1] == 24

def test_generate_new_random_trade_position_time_series_len():
    """Check that the period and volume series are of the same length"""
    new_trade = generate_new_random_trade_position(date="01/04/2015")
    period_list = new_trade["time"]
    volume_list = new_trade["volume"]

    assert len(period_list) == len(volume_list)


src/powerservice/trading.py
"""
Module that contains the get trades functionality. This module will generate a random set of dummy positions.
"""
import random
import logging
import datetime
import uuid

import numpy as np
import pandas as pd


def check_if_valid_date(date: str):
    """
    Verify that the date format matches d/m/y
    :param date: str date in d/m/y format
    :return: True or False
    """
    date_format = "%d/%m/%Y"

    """ Warning to any non python devs reading this code..
        In Python the only way to test a valid date is with a try catch. Yep, it sux.
    """
    if not isinstance(date, str):
        return False

    try:
        datetime.datetime.strptime(date, date_format)
        valid_date = True
    except ValueError:
        valid_date = False

    return valid_date


def random_nan(x):
    """
    Replace x with a nan, if the random number == 1
    """
    if random.randrange(0, 15) == 1:
        x = np.nan

    return x


def generate_new_random_trade_position(date: str):
    """ Generates a new random trade position with the date, period sequence and volume sequence
    :param date: Date in d/m/y format
    :return: dict with data
    """

    period_list = [random_nan(i.strftime("%H:%M")) for i in pd.date_range("00:00", "23:59", freq="5min").time]
    volume = [random_nan(x) for x in random.sample(range(0, 500), len(period_list))]

    open_trade_position = {"date": date,
                           "time": period_list,
                           "volume": volume,
                           "id": uuid.uuid4().hex
                           }

    return open_trade_position


def get_trades(date: str):
    """
    Generate some random number of open trade positions
    :param date: date in d/m/y format
    :return:
    """

    if not check_if_valid_date(date=date):
        error_msg = "The supplied date {} is invalid.Please supply a date in the format d/m/Y.".format(date)
        logging.error(error_msg)
        raise ValueError(error_msg)

    # a randomly chosen number of open trades
    number_of_open_trades = random.randint(1, 101)
    logging.info("Generated" + str(number_of_open_trades) + " open trades randomly.")

    open_trades_list = []
    # Generate a list of open trade dicts
    for open_trade in range(0, number_of_open_trades):
        open_trades_list.append(generate_new_random_trade_position(date=date))

    return open_trades_list


if __name__ == '__main__':
    trades = get_trades(date='01/03/2022')
    df = pd.DataFrame(trades[0])
    print(df)



data_engineering_pipeline.py
from powerservice import trading
import pandas as pd
import numpy as np
from datetime import datetime
import csv

# Importing the relevant datasets
trades_today = trading.get_trades(date='14/09/2022')
trades_yesterday = trading.get_trades(date='13/09/2022')

initial = pd.DataFrame(trades_today[0])
# print(initial)


# Aggregating the data
times = pd.to_datetime(initial.time)
aggregated = initial.groupby([times.dt.hour]).agg(date=('date', 'max'), volume=("volume", "sum"),
                                                         id=('id', 'max'))
aggregated = aggregated.reset_index()[['date', 'time', 'volume', 'id']]


# Selecting for the required columns
time_volume = aggregated[["time", "volume"]]
# print(time_volume)

# Collecting the values from yesterday
previous = pd.DataFrame(trades_yesterday[0])
# print(previous)
previous = previous.reset_index()[['date', 'time', 'volume', 'id']]
# print(previous)

# Selecting the values from 23:00
previous['time'] = pd.to_datetime(previous['time'])
start_time = previous.set_index('time').between_time('23:00', '23:59')
start_time = start_time.reset_index()[['date', 'time', 'volume', 'id']]

# Aggregating the values
times2 = pd.to_datetime(start_time.time)
start_time = start_time.groupby([times2.dt.hour]).agg(date=('date', 'max'), volume=("volume", "sum"), id=('id', 'max'))
start_time = start_time.reset_index()[['date', 'time', 'volume', 'id']]
# Selecting for the required columns
start_t_v = start_time[["time", "volume"]]

# Moving the 23:00 values to the beginning of the DataFrame
from_start = pd.concat([start_t_v, time_volume])

# Converting time elements to 24-hour format
from_start['time'] = pd.to_datetime(from_start.time, format='%H')
from_start['time'] = from_start['time'].dt.strftime('%H:%M')
# print(from_start)

# Dropping the second 23:00 value, resetting the index and deleting the old index column
from_start = from_start.drop(23)
from_start = from_start.reset_index()
from_start = from_start.drop(columns="index")
# print(from_start)

# Exporting the CSV file in the required format
now = datetime.now()
dt_string = now.strftime("%Y%m%d_%H%M")
# print(dt_string)
from_start.to_csv('PowerPosition_' + dt_string + '.csv', index=False)

# Exporting the Data Profiling CSV
data_profile = [('PowerPosition_' + dt_string + '.csv')]
with open('PowerPosition_' + dt_string + '_data_profiling' + '.csv', 'w', newline='') as csvfile:
    my_writer = csv.writer(csvfile, delimiter=' ')
    my_writer.writerow(data_profile)

file_name = 'PowerPosition_' + dt_string + '_data_quality.csv'

datetime_series = pd.to_datetime(initial['time'])
datetime_index = pd.DatetimeIndex(datetime_series.values)
df3 = initial.set_index(datetime_index)

df4 = df3.resample('5T').mean()

df3 = df3.set_index(df4.index)
df3 = df3.set_index(df4.index).reset_index()
df3 = df3.fillna(0)
df_quality1 = df3[df3['time'] == 0]
df_quality1['index'] = df_quality1['index'].dt.strftime('%H:%M')

df_quality1.rename(columns={'index': 'missed_intervals'}, inplace=True)
df_quality1 = df_quality1[["date", "missed_intervals", "id"]]

fields = ['<< TIME INTERVAL CHECK >>']
with open(file_name, 'a') as f:
    writer = csv.writer(f)
    writer.writerow(fields)

# Exporting Data Quality CSV
df_quality1.to_csv(file_name, mode='a', index=False)

df_quality2 = df3[(df3['volume'] == 0) & (df3['time'] != 0)]
df_quality2 = df_quality2[["date", "time", "volume", "id"]]

fields = ['<< MISSING VALUES CHECK >>']
with open(file_name, 'a') as f:
    writer = csv.writer(f)
    writer.writerow(fields)

# Exporting Data Quality CSV
df_quality2.to_csv(file_name, mode='a', index=False)

if from_start.time[0] == '23:00' and from_start.time[23] == '22:00':
    fields = ['<< START AND END TIMES ARE VALID >>']
else:
    fields = ['<< START AND END TIMES ARE INVALID >>']
with open(file_name, 'a') as f:
    writer = csv.writer(f)
    writer.writerow(fields)

if pd.to_datetime(from_start['time'], format='%H:%M', errors='coerce').notnull().all():
    fields = ['<< TIME COLUMN VALUES HAVE VALID FORMAT >>']
else:
    fields = ['<< TIME COLUMN VALUES HAVE INVALID FORMAT >>']
with open(file_name, 'a') as f:
    writer = csv.writer(f)
    writer.writerow(fields)


data_engineering_pipeline_BD.py
from powerservice import trading
import pandas as pd
import numpy as np
from datetime import datetime
import csv

# Importing the relevant datasets
trades_today = trading.get_trades(date='14/09/2022')
trades_yesterday = trading.get_trades(date='13/09/2022')

initial = pd.DataFrame(trades_today)
# print(initial)

# Flattening the nested values within the time and volume columns
flat_initial = initial.explode(['time', 'volume'])
# print(flat_initial.head(40))
flat_initial = flat_initial.reset_index()[['date', 'time', 'volume', 'id']]
# print(flat_initial)
flat_initial["volume"] = flat_initial.volume.astype(float)
flat_initial1 = flat_initial
# print(flat_initial1.head(30))


# Aggregating the data
times = pd.to_datetime(flat_initial.time)
flat_initial = flat_initial.groupby([times.dt.hour]).agg(date=('date', 'max'), volume=("volume", "sum"),
                                                         id=('id', 'max'))
flat_initial = flat_initial.reset_index()[['date', 'time', 'volume', 'id']]
# print(flat_initial)

# Selecting for the required columns
time_volume = flat_initial[["time", "volume"]]
# print(time_volume)

# Collecting the values from yesterday
previous = pd.DataFrame(trades_yesterday)
# Flattening the nested values within the time and volume columns
previous = previous.explode(['time', 'volume'])
# print(previous)
previous = previous.reset_index()[['date', 'time', 'volume', 'id']]
# print(previous)

# Selecting the values from 23:00
previous['time'] = pd.to_datetime(previous['time'])
start_time = previous.set_index('time').between_time('23:00', '23:59')
start_time = start_time.reset_index()[['date', 'time', 'volume', 'id']]

# Aggregating the values
times2 = pd.to_datetime(start_time.time)
start_time = start_time.groupby([times2.dt.hour]).agg(date=('date', 'max'), volume=("volume", "sum"), id=('id', 'max'))
start_time = start_time.reset_index()[['date', 'time', 'volume', 'id']]
# Selecting for the required columns
start_t_v = start_time[["time", "volume"]]

# Moving the 23:00 values to the beginning of the DataFrame
from_start = pd.concat([start_t_v, time_volume])
# print(from_start)

# Converting time elements to 24-hour format
from_start['time'] = pd.to_datetime(from_start.time, format='%H')
from_start['time'] = from_start['time'].dt.strftime('%H:%M')
# print(from_start)

# Dropping the second 23:00 value, resetting the index and deleting the old index column
from_start = from_start.drop(23)
from_start = from_start.reset_index()
from_start = from_start.drop(columns="index")
# print(from_start)

# Exporting the CSV file in the required format
now = datetime.now()
dt_string = now.strftime("%Y%m%d_%H%M")
# print(dt_string)
from_start.to_csv('PowerPosition_' + dt_string + '.csv', index=False)

# Exporting the Data Profiling CSV
data_profile = [('PowerPosition_' + dt_string + '.csv')]
with open('PowerPosition_' + dt_string + '_data_profiling' + '.csv', 'w', newline='') as csvfile:
    my_writer = csv.writer(csvfile, delimiter=' ')
    my_writer.writerow(data_profile)

file_name = 'PowerPosition_' + dt_string + '_data_quality.csv'

if from_start.time[0] == '23:00' and from_start.time[23] == '22:00':
    fields = ['<< START AND END TIMES ARE VALID >>']
else:
    fields = ['<< START AND END TIMES ARE INVALID >>']
with open(file_name, 'a') as f:
    writer = csv.writer(f)
    writer.writerow(fields)

if pd.to_datetime(from_start['time'], format='%H:%M', errors='coerce').notnull().all():
    fields = ['<< TIME COLUMN VALUES HAVE VALID FORMAT >>']
else:
    fields = ['<< TIME COLUMN VALUES HAVE INVALID FORMAT >>']
with open(file_name, 'a') as f:
    writer = csv.writer(f)
    writer.writerow(fields)

flat_initial1['time'] = pd.to_datetime(flat_initial1['time'])
flat_initial1 = flat_initial1.set_index('time')

df1 = (flat_initial1.groupby([pd.Grouper(freq='D'), 'id', 'date'])['volume']
       .apply(lambda x: x.asfreq('5min'))
       .reset_index(level=0, drop=True)
       .reset_index())
df1 = df1.reset_index()[['date', 'time', 'volume', 'id']]
df1['time'] = df1['time'].dt.strftime('%H:%M')
df2 = df1[df1['volume'].isna()]
df3 = df1.fillna('0')

fields = ['<< MISSING TIME INTERVALS: >>']
with open(file_name, 'a') as f:
    writer = csv.writer(f)
    writer.writerow(fields)

# export csv file
df2.to_csv(file_name, mode='a', index=False)

df_quality = df3[(df3['volume'] == 0) & (df3['time'] != 0)]
df_quality = df_quality[["date", "time", "volume", "id"]]

fields = ['<< MISSING VALUES CHECK >>']
with open(file_name, 'a') as f:
    writer = csv.writer(f)
    writer.writerow(fields)

# Exporting Data Quality CSV
df_quality.to_csv(file_name, mode='a', index=False)
